staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -696,6 +696,7 @@ func (w *watchCache) isIndexValidLocked(index int) bool {
 // be called under the watchCache lock.
 func (w *watchCache) getAllEventsSinceLocked(resourceVersion uint64, key string, opts storage.ListOptions) (*watchCacheInterval, error) {
 	_, matchesSingle := opts.Predicate.MatchesSingle()
+	matchesSingle = matchesSingle && !opts.Recursive
```

Comment:
I think we can be smarter here by using a different key here, but let's fix that first.

Reply:
yeah, we would have to construct the full key. is that what you had in mind as well ?

Reply:
Yes - but we should do that as a follow-up (if at all)